### PR TITLE
fix: negotiate handshake until the end in wolfSSL_read/wolfSSL_write

### PR DESCRIPTION
**Pull request by rizlik**
# Description

when checking if we need to negotiate do not rely on `ssl->options.handShakeState == HANDSHAKE_DONE` because is set too early and it doesn't account for I/O non-blocking errors. 

If the server has a Finished message in the output buffer, it will not re-enter
in `wolfSSL_negotiate()`, because the handshake is completed according to `ssl->options.handShakeState == HANDSHAKE_DONE` state machine and this will cause the handshake to stall.

# Testing

Added a negative test for this particular edge case.

#ZD 17433

### CODE REVIEW

**Review by julek-wolfssl — DISMISSED** [review]
Great find. Please add a test with the write/read going the other way.

**Comment by SparkiDev** [comment]
Application should be checking wolfSSL_want_write().
Is this not going to work for customer?

**Comment by rizlik** [comment]
> Application should be checking wolfSSL_want_write(). Is this not going to work for customer?

I don't think it will solve the problem:
 If the application does the handshake using only wolfSSL_read/write() instead of wolfSSL_accept/connect part of the accept/connect state machine may be skipped.  The issue is that receive/send_data don't enter negotiate because of `ssl->options.handShakeDone == 1`, set before the full connect/accept state machine completes. For example, besides not flushing the output buffer, the `ssl->hsDoneCb` is not invoked if, on the server side, the read/write function returns just after building the FINISHED message (for example bc of WANT_WRITE). I'm working on an alternative solution as the approach of this PR can't work.

**Comment by rizlik** [comment]
retest this

File: src/internal.c [hunk 1]
```diff
@@ -24040,6 +24040,51 @@ static int CheckTLS13AEADSendLimit(WOLFSSL* ssl)
...
+/*
+ * Invoked in wolfSSL_read/wolfSSL_write to check if wolfSSL_negotiate() is
+ * needed in the handshake.
+ *
+ * In TLSv1.2 negotiate until the end of the handshake, unless:
+ * 1 in SCR and sending data or
+ * 2 in SCR and we have plain data ready
+ * Early data logic may bypass this logic in TLSv1.3 when appropriate.
+ */
+static int need_negotiate(WOLFSSL *ssl, int send)
```
Comment (SparkiDev) [inline src/internal.c:24052]: Rename to: ssl_in_handshake or ssl_handshake_done
Or something similar.

File: src/internal.c [hunk 2]
```diff
@@ -24040,6 +24040,51 @@ static int CheckTLS13AEADSendLimit(WOLFSSL* ssl)
...
+            return 0;
+        } else {
+            /* allow reading buffered data in SCR */
+            if (ssl->buffers.clearOutputBuffer.length != 0)
+                return 0;
+        }
+        return 1;
+    }
+
+    if (ssl->options.handShakeState != HANDSHAKE_DONE)
```
Comment (SparkiDev) [inline src/internal.c:24066]: Does ssl->options.handShakeDone help?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Comment by matteofumagalli1275** [comment]
Hello, i'm the creator of the issue at #ZD 17433

Today i'm gonna test the fix and let you know

File: src/internal.c (hunk 2 above)
Comment (rizlik) [reply src/internal.c:24066]: not really. It's set at the same time of `ssl->options.handShakeState` and it's not re-set on renegotiation.

**Comment by matteofumagalli1275** [comment]
I can confirm that with the fix negotiation works correctly in my test cases even without calling accept/connect.
Tested on https://github.com/wolfSSL/wolfssl/pull/7237/commits/a7016b400c9835d1ee8b29a88954485ca3818025

**Comment by rizlik** [comment]
retest this please

**Review by SparkiDev — APPROVED** [review]

**Review by dgarske — APPROVED** [review]
